### PR TITLE
Modify pyTorch Data Loader to Match TensorFlow Output

### PR DESCRIPTION
**Pull request by RexxarCHL**

### CODE REVIEW

File: costar_dataset/block_stacking_reader_torch.py [hunk 1]
```diff
@@ -911,23 +924,26 @@ def ConvertImageListToNumpy(data, format='numpy', data_format='NHWC'):
     output_shape = (224, 224, 3)
     # output_shape = None
     # tf.enable_eager_execution()
-    filenames = glob.glob(os.path.expanduser('~/.keras/datasets/costar_block_stacking_dataset_v0.4/blocks_only/*success.h5f'))
+    filenames = glob.glob(os.path.expanduser('~/Documents/costar_block_stacking_dataset_v0.4/blocks_only/*success.h5f'))
```
Comment (ahundt) [inline costar_dataset/block_stacking_reader_torch.py]: could you revert this line?

**Comment by ahundt** [comment]
almost LGTM, just one line to revert

**Comment by ahundt** [comment]
Oh wait, can this use the train/val/test list splits?

**Review by ahundt — DISMISSED** [review]
almost LGTM, see comments already made

**Comment by RexxarCHL** [comment]
Path reverted.

File: costar_dataset/block_stacking_reader_torch.py [hunk 2]
```diff
@@ -616,6 +622,68 @@ def __init__(self, list_example_filenames,
...
+                        total_actions_available=41,
+                        seed=0, random_state=None,
+                        is_training=True, random_augmentation=None,
+                        random_shift=False,
+                        output_shape=None,
+                        blend_previous_goal_images=False,
+                        estimated_time_steps_per_example=250, verbose=0, inference_mode=False, one_hot_encoding=True,
+                        pose_name='pose_gripper_center',
+                        force_random_training_pose_augmentation=None):
+        '''Call constructor from specified parameter
```
Comment (ahundt) [inline costar_dataset/block_stacking_reader_torch.py]: Can you add a proper docstring to this? What is it for, what do you integrate it with, why should you call it? Where can I find a useful example?

File: costar_dataset/block_stacking_reader_torch.py [hunk 3]
```diff
@@ -705,7 +773,7 @@ def ConvertImageListToNumpy(data, format='numpy', data_format='NHWC'):
                     raise ValueError('CostarBlockStackingDataset: Trying to open something which is not a file: ' + str(example_filename))
                 with h5py.File(example_filename, 'r') as data:
                     if 'gripper_action_goal_idx' not in data or 'gripper_action_label' not in data:
-                        raise ValueError('block_stacking_reader.py: You need to run preprocessing before this will work! \n' +
+                        raise ValueError('CostarBlockStackingDataset: You need to run preprocessing before this will work! \n' +
                                          '    python2 ctp_integration/scripts/view_convert_dataset.py --path ~/.keras/datasets/costar_block_stacking_dataset_v0.4 --preprocess_inplace gripper_action --write'
```
Comment (ahundt) [inline costar_dataset/block_stacking_reader_torch.py:805]: hmm, do we have view_convert_dataset in the costar dataset repository? Is it independent enough that we could move it here?

File: costar_dataset/block_stacking_reader_torch.py (hunk 2 above)
Comment (ahundt) [reply costar_dataset/block_stacking_reader_torch.py]: Also, what are the options for stuff like feature_mode

**Review by ahundt — CHANGES_REQUESTED** [review]
Looks like great progress! just a few points of confusion

File: costar_dataset/block_stacking_reader_torch.py (hunk 2 above)
Comment (RexxarCHL) [reply costar_dataset/block_stacking_reader_torch.py]: Detailed docstring added.

File: costar_dataset/block_stacking_reader_torch.py (hunk 3 above)
Comment (RexxarCHL) [reply costar_dataset/block_stacking_reader_torch.py:805]: I don't think view_convert_dataset is really needed in the costar_dataset repo. 
We will most likely run the preprocessing script before we upload the dataset, so this situation will likely never happen.
We could move the script in, but it would add additional dependencies for the package.

File: costar_dataset/__init__.py [hunk 4]
```diff
@@ -2,4 +2,21 @@
 
 __version__ = '0.4.0'
 __author__ = 'Andrew Hundt <ATHundt@gmail.com>'
-__all__ = []
+
+try:
+    from costar_dataset.block_stacking_reader_torch import CostarBlockStackingDataset
+    from costar_dataset.hypertree_pose_metrics_torch import cart_error, angle_error, grasp_acc_in_bins_batch
+except ImportError:
+    print("Torch not installed.")
```
Comment (ahundt) [inline costar_dataset/__init__.py]: This print statement doesn't seem true... why would it only error on torch and not other missing libraries?

File: costar_dataset/block_stacking_reader_torch.py [hunk 5]
```diff
@@ -616,6 +623,91 @@ def __init__(self, list_example_filenames,
...
+                        seed=0, random_state=None,
+                        is_training=True, random_augmentation=None,
+                        random_shift=False,
+                        output_shape=None,
+                        blend_previous_goal_images=False,
+                        estimated_time_steps_per_example=250, verbose=0, inference_mode=False, one_hot_encoding=True,
+                        pose_name='pose_gripper_center',
+                        force_random_training_pose_augmentation=None):
+        '''
+        Calls the class constructor from specified parameter.
```
Comment (ahundt) [inline costar_dataset/block_stacking_reader_torch.py]: This sentence has no useful meaning, every function has parameters, could you rephrase?

File: costar_dataset/block_stacking_reader_torch.py [hunk 6]
```diff
@@ -616,6 +623,91 @@ def __init__(self, list_example_filenames,
...
+        This function saves the hussle and redundant code of opening and reading respective filename txt for different dataset splits.
```
Comment (ahundt) [inline costar_dataset/block_stacking_reader_torch.py]: respective filename txt for different dataset spits? What does this sentence mean, please rephrase

Comment (ahundt) [reply costar_dataset/block_stacking_reader_torch.py]: The best thing to do generally for docstrings is to think about what someone might think when they've never seen this code before. When would you want to call this? What kind of parameter? Function parameter? command line parameter? why do you want to set parameters at all?

Generally, words like "parameter" and "variable" are useless since they can be used in every single programming context. Specifically explaining what/where/ and why is key.

File: costar_dataset/block_stacking_reader_torch.py [hunk 7]
```diff
@@ -885,6 +979,19 @@ def ConvertImageListToNumpy(data, format='numpy', data_format='NHWC'):
                 raise ValueError('Unsupported input data for y: ' + str(x))
 
             # Assemble the data batch
+            # HACK(rexxarchl): Squeeze to match the output from tensorflow
```
Comment (ahundt) [inline costar_dataset/block_stacking_reader_torch.py]: how do the outputs differ? The answer to that question might be useful to specify in the docstring, & include a link to the relevant pytorch/tf docs

File: setup.py [hunk 8]
```diff
@@ -38,7 +38,11 @@ def read(filename):
         'pyquaternion',
         'scikit-learn',
         'h5py',
-        'tqdm'],
+        'tqdm',
+        'PIL',  # TODO(rexxarchl): or pillow?
```
Comment (ahundt) [inline setup.py]: definitely pillow

**Review by ahundt — APPROVED** [review]
LGTM, thanks!
